vllm/v1/core/kv_cache_manager.py

```diff
@@ -120,47 +123,45 @@ def append_slots(
             # slots, but we cannot allocate new blocks due to the limit.
             return None
 
-        # When caching is enabled, assign token IDs to already allocated blocks.
-        new_token_ids = None
-        parent_block = None
-        if self.enable_caching:
-            # Figure out the token IDs to add to the blocks.
-            new_token_ids = request.all_token_ids[
-                request.num_computed_tokens:request.num_computed_tokens +
-                num_tokens]
-
-            # Find the last full block index.
-            # TODO: This may be optimized by calculating the computed tokens.
-            last_full_block_idx = len(req_blocks) - 1
-            while (last_full_block_idx >= 0
-                   and req_blocks[last_full_block_idx].block_hash is None):
-                last_full_block_idx -= 1
-
-            parent_block = (req_blocks[last_full_block_idx]
-                            if last_full_block_idx >= 0 else None)
-            token_id_idx = self._add_token_ids_to_blocks(
-                blocks=req_blocks[last_full_block_idx + 1:],
-                token_ids=new_token_ids,
-                parent_block=parent_block)
-
-            new_token_ids = new_token_ids[token_id_idx:]
-            parent_block = req_blocks[-1]
-
-        # No new block is needed. When caching is enabled, we make sure
-        # token_id_idx is equal to len(new_token_ids), meaning that all tokens
-        # are added to allocated blocks.
+        # No new block is needed.
         if num_required_blocks <= len(req_blocks):
-            assert not self.enable_caching or token_id_idx == num_tokens, \
-                    f"{token_id_idx=} != {num_tokens=}"
             return []
 
-        # Allocate new blocks considering preallocated blocks, and
-        # add token IDs to them if caching is enabled.
-        num_new_blocks = min(num_new_blocks + self.num_preallocate_blocks,
-                             self.free_block_queue.num_free_blocks)
-        new_blocks = self._get_new_blocks(num_new_blocks, new_token_ids,
-                                          parent_block)
+        # Get new blocks from the free block pool.
+        num_new_blocks = min(
+            num_new_blocks + self.num_preallocate_blocks,
+            self.free_block_queue.num_free_blocks,
+        )
+
+        new_blocks = self._get_new_blocks(num_new_blocks)
         req_blocks.extend(new_blocks)
+
+        if not self.enable_caching:
+            return new_blocks
+
+        num_computed_full_blocks = (request.num_computed_tokens //
+                                    self.block_size)
+
+        # NOTE(rickyx): We are assuming the `num_tokens` are actual
+        # tokens rather than placeholders for lookahead slots. If not
```

Comment:
What placeholders are you referring to? Is it image token or the speculative decoding slots?

Reply:
i was referring to speculative decoding slots - lets say when we support it, we will have X number of lookahead tokens, but these tokens shouldn't make any full block containing it hashable.

Reply:
Got it. Please make it explicitly as we haven't supported speculative decoding in v1. It would be good to let people who are going to implement it be aware of this.